samples/snippets/powershell/VS_Snippets_CLR_System/system.Collections.Hashtable_ClassExample/PS/hashtable_example.ps1

```diff
@@ -0,0 +1,116 @@
+# <snippet00>
+# Create new hash table
+# Create as ordered
+$OpenWith = @{}
+
+# Add some elements to the hash table. There are no
+# duplicate keys but some of the values are duplicated
+$OpenWith.Add('txt', 'notepad.exe')
+$openWith.Add('dib', 'paint.exe')
+
+# Add in a PowerShell way!
+$OpenWith['bmp'] = 'paint.exe'
+$OpenWith.rtf    = 'wordpad.exe'
+
+# The Add method throws an exception if the new key is 
+# already in the hash table.
+try
+{
+  $OpenWith.Add("txt", "winword.exe")
+}
+catch
+{
+  "An element with Key = 'txt' already exists."
+}
+ 
+# The Item property is the default property, so you 
+# can omit its name when accessing elements. 
+"For key = 'rtf', value = {0}." -f $OpenWith["rtf"]
+
+# The default Item property can be used to change the value
+# associated with a key.
+$OpenWith.rtf = "winword.exe";
+"For key = 'rtf', value = {0}." -f $openWith["rtf"]
+        
+# If a key does not exist, setting the default Item property
+# for that key adds a new key/value pair.
+$OpenWith.doc = "winword.exe";
+
+# ContainsKey can be used to test keys before inserting 
+# them.
+if (-Not $openWith.ContainsKey("ht"))
+{
+   $OpenWith.Add("ht", "hypertrm.exe")
+   "Value added for key = 'ht' [{0}]" -f $openWith["ht"]
+}
+
+# When you use foreach to enumerate hash table elements,
+# the elements are retrieved as KeyValuePair objects.
+""
+foreach($de in ($OpenWith.getenumerator() ) )
+{
+  "Key = {0}, Value = {1}" -f $de.Key, $de.Value
+}
+
+# To get the values alone, use the Values property.
+$ValueColl = $OpenWith.Values
+
+# The elements of the ValueCollection are strongly typed
```

Comment:
I think the context of "strongly typed" matters more in C# than it does in PowerShell but I lack the expertise to say so definitively.

Reply:
Let's be honest, it's totally meaningless to say that things are "strongly typed" in a hashtable -- they are boxed to `System.Object`  and could be (literally) _anything_...

Reply:
I am not sure what you mean @Jaykul 

The value in a hash table entry can indeed be strongly typed - as shown

PSH [C:\foo]: $a=@{}
PSH [C:\foo]: $a.x = [int64] 12345
PSH [C:\foo]: $a.GetEnumerator() | foreach {$_.value | gm}
   TypeName: System.Int64
...

Reply:
I guess I'm just quibbling about your terminology. It's not "strongly" typed. Strongly typed is when it enforces strict restrictions on the type ... hashtables don't enforce any type restrictions at all.

### This is strongly typed:

```
PSH [C:\foo]: $a = [System.Collections.Generic.Dictionary[String, Int64]]@{}
PSH [C:\foo]: $a.x = 2
PSH [C:\foo]: $a.x.GetType().FullName
System.Int64
PSH [C:\foo]: $a.x = "Hello"
Cannot convert value "Hello" to type "System.Int64". Error: "Input string was not in a correct format."
```

### This is not:

```
PSH [C:\foo]: $a=@{}
PSH [C:\foo]: $a.x = [int64] 12345
PSH [C:\foo]: $a.x.GetType().FullName
System.Int64
PSH [C:\foo]: $a.x = "Hello"
PSH [C:\foo]: $a.x.GetType().FullName
System.String
```

In the case of hashtables, the Int64 has to be up-cast to `Object` to store it, regardless of the language.

However, _unlike C#_, PowerShell **always** treats objects as their real base type. That is, it simply ignores up-casting.  That's a core feature of the language (and it's not specific to hashtables). Unlike in C#, you never have to down-cast an `Object` to use it.

It's probably worth keeping an example to point that out:

PowerShell is a dynamic language, so it ignores up-casting. You can always treat the values in a hashtable as if they were the original object you put in there, rather than being up-cast to `Object`. This means (for instance) that if you put a number into the hashtable, you can do math directly on it:

```
PSH [C:\foo]: $a=@{}
PSH [C:\foo]: $a.x = [int64] 12345
PSH [C:\foo]: $a.x *= 2
PSH [C:\foo]: $a.x
24690
```

Reply:
thanks, @Jaykul, for the definitive answer i was looking for :)